File: Design Patterns/Behavioral/TemplateMethod.ts
abstract class OrderProcessor {
    processOrder() {
        this.verifyOrder()
        this.processPayment()
        this.packItems()
        this.shipOrder()
        this.sendConfirmation()
    }

    verifyOrder() { console.log("Order verified") }

    abstract processPayment(): void
    abstract packItems(): void
    abstract shipOrder(): void

    sendConfirmation() { console.log("Confirmation email sent") }
}

class StandardOrder extends OrderProcessor {
    processPayment() { console.log("Processing credit card payment") }

    packItems() { console.log("Packing physical items") }

    shipOrder() { console.log("Shipping via standard delivery") }
}

class DigitalOrder extends OrderProcessor {
    processPayment() { console.log("Processing online payment") }

    packItems() { console.log("No packaging required") }

    shipOrder() { console.log("Sending download link") }
}

const order = new StandardOrder()
order.processOrder()

File: Design Patterns/Behavioral/C#/TemplateMethod.cs
using System;
abstract class OrderProcessor
{
    public void ProcessOrder()
    {
        this.VerifyOrder();
        this.ProcessPayment();
        this.PackItems();
        this.ShipOrder();
        this.SendConfirmation();
    }

    protected void VerifyOrder() { Console.WriteLine("Order verified"); }

    protected abstract void ProcessPayment();
    protected abstract void PackItems();
    protected abstract void ShipOrder();
    protected void SendConfirmation() { Console.WriteLine("Confirmation email sent"); }
}

class StandardOrder : OrderProcessor
{
    protected override void ProcessPayment() { Console.WriteLine("Processing credit card payment"); }
    protected override void PackItems() { Console.WriteLine("Packing physical items"); }
    protected override void ShipOrder() { Console.WriteLine("Shipping via standard delivery"); }
}

class DigitalOrder : OrderProcessor
{
    protected override void ProcessPayment() { Console.WriteLine("Processing online payment"); }
    protected override void PackItems() { Console.WriteLine("No packaging required"); }
    protected override void ShipOrder() { Console.WriteLine("Sending download link"); }
}

class Program
{
    public static void Main(string[] args)
    {
        StandardOrder order = new StandardOrder();
        order.ProcessOrder();
    }
}